File: rusty_c/mydisk/src/TreediskFS_RS/clock.py
class Block:
  def __init__(self, data: list[bytes], ref_bit: bool, inode, offset, is_dirty):
    self.data = data 
    self.ref_bit = ref_bit
    self.inode = inode
    self.offset = offset
    self.is_dirty = is_dirty

class ClockCache:
  def __init__(self, capacity):
    self.capacity = capacity
    self.len = 0
    self.arr = [None] * capacity
    self.clock_hand = 0
    # lookup[(node, offset)] = idx
    self.lookup = {}

  def clock_find(self) -> int:
    """Find the next block to evict using the clock algorithm
    Assumes that the cache is full."""
    if self.arr[self.clock_hand].ref_bit:
      self.arr[self.clock_hand].ref_bit = False
      self.clock_hand = (self.clock_hand + 1) % self.capacity
      return self.clock_find()
    else:
      return self.clock_hand

  def read_below(self, inode, offset) -> list[bytes]:
    return Block([b'0x0'] * 512, True, inode, offset, False)
  
  def write_below(self, inode, offset, data: list[bytes]):
    pass

  def read(self, inode, offset) -> list[bytes]:
    if (inode, offset) in self.lookup:
      idx = self.lookup[(inode, offset)]
      self.arr[idx].ref_bit = True
      self.clock_hand = idx
      return self.arr[idx].data
    else:
      if self.len < self.capacity:
        self.arr[self.len] = self.read_below(inode, offset)
        self.lookup[(inode, offset)] = self.len
        self.arr[self.len].ref_bit = True
        self.arr[self.len].is_dirty = False
        self.clock_hand = self.len
        self.len += 1
        return self.arr[self.len - 1].data
      else:
        idx = self.clock_find()
        block_to_evict = self.arr[idx]
        if block_to_evict.is_dirty:
          self.write_below(block_to_evict.inode, block_to_evict.offset, block_to_evict.data)
        self.arr[idx] = self.read_below(inode, offset)
        self.lookup[(inode, offset)] = idx
        self.arr[idx].ref_bit = True
        self.clock_hand = idx
        self.arr[idx].is_dirty = False
        del self.lookup[(block_to_evict.inode, block_to_evict.offset)]
        return self.arr[idx].data
      
  def write(self, inode, offset, data):
    if (inode, offset) in self.lookup:
      idx = self.lookup[(inode, offset)]
      self.arr[idx] = Block(data, True, inode, offset, True)
      self.clock_hand = idx
    else:
      if self.len < self.capacity:
        self.arr[self.len] = Block(data, True, inode, offset, True)
        self.lookup[(inode, offset)] = self.len
        self.clock_hand = self.len
        self.len += 1
      else:
        idx = self.clock_find()
        block_to_evict = self.arr[idx]
        if block_to_evict.is_dirty:
          self.write_below(block_to_evict.inode, block_to_evict.offset, block_to_evict.data)
        self.arr[idx] = Block(data, True, inode, offset, True)
        self.lookup[(inode, offset)] = idx
        self.clock_hand = idx
        del self.lookup[(block_to_evict.inode, block_to_evict.offset)]

File: rusty_c/mydisk/src/TreediskFS_RS/clock.rs
use std::collections::HashMap;

struct Below {
    mp: HashMap<(i32, i32), Vec<u8>>,
}

impl Below {
    fn new() -> Self {
        Below {
            mp: HashMap::new(),
        }
    }

    fn read(&mut self, inode: i32, offset: i32) -> Option<&Vec<u8>> {
        self.mp.get(&(inode, offset))
    }

    fn write(&mut self, inode: i32, offset: i32, data: Vec<u8>) {
        self.mp.insert((inode, offset), data);
    }
}

struct Block {
    data: Vec<u8>,
    ref_bit: bool,
    inode: i32,
    offset: i32,
    is_dirty: bool,
}

struct ClockCache {
    capacity: usize,
    len: usize,
    arr: Vec<Option<Block>>,
    clock_hand: usize,
    lookup: HashMap<(i32, i32), usize>,
}

impl ClockCache {
    fn new(capacity: usize) -> Self {
        ClockCache {
            capacity,
            len: 0,
            arr: vec![None; capacity],
            clock_hand: 0,
            lookup: HashMap::new(),
        }
    }

    fn clock_find(&mut self) -> usize {
        while self.arr[self.clock_hand].as_ref().unwrap().ref_bit {
            self.arr[self.clock_hand].as_mut().unwrap().ref_bit = false;
            self.clock_hand = (self.clock_hand + 1) % self.capacity;
        }
        self.clock_hand
    }

    fn read_below(&mut self, below: &mut Below, inode: i32, offset: i32) -> Block {
        let data = below.read(inode, offset).cloned().unwrap_or_else(Vec::new);
        Block {
            data,
            ref_bit: true,
            inode,
            offset,
            is_dirty: false,
        }
    }

    fn write_below(&mut self, below: &mut Below, inode: i32, offset: i32, data: Vec<u8>) {
        below.write(inode, offset, data);
    }

    fn read(&mut self, below: &mut Below, inode: i32, offset: i32) -> Vec<u8> {
        if let Some(&idx) = self.lookup.get(&(inode, offset)) {
            self.arr[idx].as_mut().unwrap().ref_bit = true;
            self.clock_hand = idx;
            self.arr[idx].as_ref().unwrap().data.clone()
        } else {
            if self.len < self.capacity {
                let block = self.read_below(below, inode, offset);
                let idx = self.len;
                self.arr[idx] = Some(block);
                self.lookup.insert((inode, offset), idx);
                self.arr[idx].as_mut().unwrap().ref_bit = true;
                self.arr[idx].as_mut().unwrap().is_dirty = false;
                self.clock_hand = idx;
                self.len += 1;
                self.arr[idx].as_ref().unwrap().data.clone()
            } else {
                let idx = self.clock_find();
                if let Some(block_to_evict) = self.arr[idx].take() {
                    if block_to_evict.is_dirty {
                        self.write_below(below, block_to_evict.inode, block_to_evict.offset, block_to_evict.data);
                    }
                }
                let block = self.read_below(below, inode, offset);
                self.arr[idx] = Some(block);
                self.lookup.insert((inode, offset), idx);
                self.arr[idx].as_mut().unwrap().ref_bit = true;
                self.clock_hand = idx;
                self.arr[idx].as_mut().unwrap().is_dirty = false;
                self.lookup.remove(&(block_to_evict.inode, block_to_evict.offset));
                self.arr[idx].as_ref().unwrap().data.clone()
            }
        }
    }

    fn write(&mut self, below: &mut Below, inode: i32, offset: i32, data: Vec<u8>) {
        if let Some(&idx) = self.lookup.get(&(inode, offset)) {
            self.arr[idx] = Some(Block {
                data,
                ref_bit: true,
                inode,
                offset,
                is_dirty: true,
            });
            self.clock_hand = idx;
        } else {
            if self.len < self.capacity {
                let block = Block {
                    data,
                    ref_bit: true,
                    inode,
                    offset,
                    is_dirty: true,
                };
                let idx = self.len;
                self.arr[idx] = Some(block);
                self.lookup.insert((inode, offset), idx);
                self.clock_hand = idx;
                self.len += 1;
            } else {
                let idx = self.clock_find();
                if let Some(block_to_evict) = self.arr[idx].take() {
                    if block_to_evict.is_dirty {
                        self.write_below(below, block_to_evict.inode, block_to_evict.offset, block_to_evict.data);
                    }
                }
                let block = Block {
                    data,
                    ref_bit: true,
                    inode,
                    offset,
                    is_dirty: true,
                };
                self.arr[idx] = Some(block);
                self.lookup.insert((inode, offset), idx);
                self.clock_hand = idx;
                self.lookup.remove(&(block_to_evict.inode, block_to_evict.offset));
            }
        }
    }

    fn below_get_size(&mut self, inode: i32) -> usize {
        0 // Replace with your implementation
    }

    fn get_size(&mut self, inode: i32) -> usize {
        self.below_get_size(inode)
    }

    fn below_set_size(&mut self, inode: i32, size: usize) {
        // Replace with your implementation
    }

    fn set_size(&mut self, inode: i32, size: usize) {
        self.below_set_size(inode, size);
    }

    fn synch(&mut self, below: &mut Below, inode: i32) {
        if inode == -1 {
            for block_idx in 0..self.len {
                if let Some(block) = self.arr[block_idx].take() {
                    if block.is_dirty {
                        self.write_below(below, block.inode, block.offset, block.data);
                    }
                }
            }
        } else {
            for block_idx in 0..self.len {
                if let Some(block) = self.arr[block_idx].take() {
                    if block.is_dirty && block.inode == inode {
                        self.write_below(below, block.inode, block.offset, block.data);
                    }
                }
            }
        }
    }
}
